Publish 24.0.0.12-beta post (#4148)
OpenLiberty/blogs: posts/2024-11-19-24.0.0.12-beta.adoc

See also link:{url-prefix}/blog/?search=beta&key=tag[previous Open Liberty beta blog posts].

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // //
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/29770
// Contact/Reviewer: neuwerk
// // // // // // // //
[#xmlws]
== Fine-tuning XML Web Services inbound SOAP message validation

Open Liberty's XML Web Services features now support fine-grained message validation for inbound SOAP messages. This enhancement provides more control over message validation options.
In Open Liberty 24.0.0.12-beta, you can configure message validation using new attributes in the `server.xml` file. These attributes are available for the `webService` and `webServiceClient` elements.

[options="header"]
|=======================
| Attribute | Description
| `enableSchemaValidation` | Enable full validation against the XML schema
| `enableDefaultValidation` | Enable or disable default validation for JAXB
| `ignoreUnexpectedElements` | Use default validation while ignoring `UnmarshallExceptions: Unknown Element` errors
|=======================

> The default value for all three attributes is `false`, except for `enableDefaultValidation` for web service clients, which defaults to `true`.

Reviewer: ```suggestion
The default value for `enableDefaultValidation` in the `webServiceClient` element is `true`. The rest of the attributes in both the `webServiceClient` and `webService` elements default to `false`.
```
I'm not sure my suggestion is that much better. I understand the sentence now, but originally it came across as "all three things are `false`, except for one that's `true`", which confused me why it would be phrased that way. I hoped my suggested change a bit more subtly implies that while there are three attributes, they apply to two different elements and therefore there are _six_ unique attributes being talked about here.

Author: I agree this is more clear- addressed in #4151

> These attributes require the one of the following XML Web Services features to be enabled in your `server.xml` file:

Reviewer: ```suggestion
These attributes require one of the following XML Web Services features to be enabled in your `server.xml` file:
```

Author: addressed in https://github.com/OpenLiberty/blogs/pull/4151

==== Global default validation

The following example shows how to enable default validation globally for web services in your Open Liberty runtime:

[source,xml]
----
<webService enableDefaultValidation="true"/>
----

Default validation is enabled by default for web service clients. To disable default validation globally for web service clients, set the `enableDefaultValidation="false"` attribute on the `webServiceClient` element.

[source,xml]
----
<webServiceClient enableDefaultValidation="false"/>
----

==== Targeted default validation

The following example shows how to enable default validation for a specific web service:

[source,xml]
----
<webService  portName="SayHelloService" enableDefaultValidation="true"/>
----

Default validation is enabled by default for web service clients. To disable default validation for a specific web service client, set the `enableDefaultValidation="false"` attribute on the `webServiceClient` element and use the `serviceName` attribute to specify the client service.

[source,xml]
----
<webServiceClient serviceName="<client service name>"  enableDefaultValidation="false" />
----

=== Ignore unexpected elements

> The new Open Liberty `ignoreUnexpectedElements` attribute provides the same function. Inbound SOAP messages often contain extra elements in the SOAP body when a web service is updated but the client is not. When a message contains an unknown element, Open Liberty throws a `UnmarshallingException: Unknown Element`. By enabling `ignoreUnkownEllements`, you can keep validation enabled while ignoring unknown elements.

Reviewer: > The new Open Liberty `ignoreUnexpectedElements` attribute provides the same function.

Provides the same function as what?

```suggestion
The new Open Liberty `ignoreUnexpectedElements` attribute provides the same function. Inbound SOAP messages often contain extra elements in the SOAP body when a web service is updated but the client is not. When a message contains an unknown element, Open Liberty throws a `UnmarshallingException: Unknown Element`. By enabling `ignoreUnexpectedElements`, you can keep validation enabled while ignoring unknown elements.
```

Author: "same function..." was an artifact from an earlier draft- Deleted and typo addressed in https://github.com/OpenLiberty/blogs/pull/4151

==== Global configuration

The following example shows how to ignore unexpected elements globally for web services on your Open Liberty runtime:

[source,xml]
----
<webService  ignoreUnexpectedElements="true"/>
----

To ignore unexpected elements globally for web service clients, set the `ignoreUnexpectedElements` attribute on the `webServiceClient` element.

==== Targeted configuration

The following example shows how to ignore unexpected elements for a specific web service:

[source,xml]
----
<webService  portName="SayHelloService" ignoreUnexpectedElements="true"/>
----

> To ignore unexpected elements for a specific web service client, set the same attribute on the `webServiceClient`element and use the `serviceName` attribute to specify the client service.

Reviewer: ```suggestion
To ignore unexpected elements for a specific web service client, set the same attribute on the `webServiceClient` element and use the `serviceName` attribute to specify the client service.
```

Author: addressed in https://github.com/OpenLiberty/blogs/pull/4151
